### PR TITLE
Move request_cache to MM Account.

### PR DESCRIPTION
**Pull request by brian-js**
[Review on Reviewable](https://reviewable.io/reviews/maidsafe/safe_vault/345)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @maqi

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/personas/maid_manager.rs [hunk 1]
```diff
@@ -225,22 +276,26 @@ impl MaidManager {
...
+        // Update the account
+        if let Err(error) = self.accounts
+                                .get_mut(&client_name)
+                                .ok_or(ClientError::NoSuchAccount)
+                                .and_then(|account| {
+                                    if type_tag != 0 {
+                                        try!(account.cache_request(message_id.clone(), &request));
+                                        account.put_data(DEFAULT_PAYMENT /* data.payload_size() as u64 */)
+                                    } else {
+                                        account.cache_request(message_id.clone(), &request)
```
Comment (maqi) [inline src/personas/maid_manager.rs]: put request of SD having type_tag = 0 won't be passed to SDM, does it still required to cache the request?

File: src/personas/maid_manager.rs [hunk 2]
```diff
@@ -119,26 +179,17 @@ impl MaidManager {
                               message_id: &MessageId,
                               external_error_indicator: &Vec<u8>)
                               -> Result<(), InternalError> {
-        match self.request_cache.remove(message_id) {
-            Some(client_request) => {
+        for (_, account) in self.accounts.clone().iter_mut() {
```
Comment (maqi) [inline src/personas/maid_manager.rs]: iterate through every account will be costly. can we do ```account = self.account.get_mut(request.dest.name())``` here?

File: src/error.rs [hunk 3]
```diff
@@ -29,6 +29,7 @@ pub enum ClientError {
     NoSuchData,
     DataExists,
     LowBalance,
+    DuplicateRequest,
```
Comment (maqi) [inline src/error.rs]: if this error is not being sent back to the client, shall it be defined under InternalError ?

File: src/personas/maid_manager.rs (hunk 1 above)
Comment (brian-js) [inline src/personas/maid_manager.rs]: Don't really know what your saying here Qi. SD go to SDM and logic hasn't changed, the request was being cached.

Comment (maqi) [inline src/personas/maid_manager.rs]: SD with tpye_tag=0 is only used for creating account in MM, and such SD doesn't get forwarded to SDM, which means there won't be any response.

File: src/personas/maid_manager.rs (hunk 2 above)
Comment (brian-js) [inline src/personas/maid_manager.rs]: What request are we talking about here?

Comment (maqi) [inline src/personas/maid_manager.rs]: sorry, then that shall be in vault.rs change to something like : 
```
            (&Authority::NaeManager(_),
             &Authority::ClientManager(dst),
             &ResponseContent::PutFailure{ ref id, ref external_error_indicator, .. }) => {
                self.maid_manager.handle_put_failure(routing_node, dst, id, external_error_indicator)
```
then here we can do
```
account = self.account.get_mut(dst.name())
```

File: src/personas/maid_manager.rs (hunk 1 above)
Comment (brian-js) [inline src/personas/maid_manager.rs]: That's not what's implemented currently. Are you referring to some part of the documentation?

Comment (maqi) [inline src/personas/maid_manager.rs]: Ah, sorry, my fault.
yes, you are correct.

File: src/personas/maid_manager.rs (hunk 2 above)
Comment (brian-js) [inline src/personas/maid_manager.rs]: Are we expecting the MM to be catering for a large number of client accounts at any given time.

Comment (maqi) [inline src/personas/maid_manager.rs]: probably yes, if the vault lives a long enough time.  :(

File: src/error.rs (hunk 3 above)
Comment (brian-js) [inline src/error.rs]: I wonder if we should send a failure with this error back to the client?

File: src/personas/maid_manager.rs (hunk 2 above)
Comment (brian-js) [inline src/personas/maid_manager.rs]: I think we can leave this as it is here and take a look later as a potential performance improvement.

File: src/error.rs (hunk 3 above)
Comment (maqi) [inline src/error.rs]: probably should.
then could you please put the code in to send this error back to the client?
thx

**Comment by maqi** [comment]
Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/345)*

**Pull request merged**